src/ucs/memory/rcache.c

```diff
@@ -962,21 +962,18 @@ void ucs_rcache_region_invalidate(ucs_rcache_t *rcache,
 {
     ucs_rcache_comp_entry_t *comp;
 
-    if (cb != NULL) {
-        /* Add completion entry before any region action to allow completion
-         * shot if region is destroyed */
-        ucs_spin_lock(&rcache->lock);
-        comp = (ucs_rcache_comp_entry_t*)ucs_mpool_get(&rcache->mp);
-        ucs_spin_unlock(&rcache->lock);
+    /* Completion entry should be added before region is invalidated */
+    ucs_spin_lock(&rcache->lock);
+    comp = ucs_mpool_get(&rcache->mp);
+    ucs_spin_unlock(&rcache->lock);
 
-        if (comp != NULL) {
-            comp->func = cb;
-            comp->arg  = arg;
-            ucs_list_add_tail(&region->comp_list, &comp->list);
-        } else {
-            ucs_rcache_region_error(rcache, region,
-                                    "Failed to allocate completion object");
-        }
+    if (comp != NULL) {
+        comp->func = cb;
+        comp->arg  = arg;
+        ucs_list_add_tail(&region->comp_list, &comp->list);
```

Comment:
BTW, should this be under lock? we access region->comp_list which may be shared

Reply:
when we extracting elements there is region destroying - it could not be under race (or this is error state), and here may be used any lock...
but ok, used pgt_lock